### PR TITLE
gh-75171: Fix parsing invalid email address headers starting or ending with a dot

### PR DESCRIPTION
**Pull request by tsufeki**
This PR is based on #2811, with additional fixes for the case where display-name starts with a dot.

Added checks whether or not given token is a list before inserting/removing cfws from it.

https://bugs.python.org/issue30988

* Issue: gh-75171

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Lib/email/_header_value_parser.py [hunk 1]
```diff
@@ -1744,7 +1744,10 @@ def get_name_addr(value):
             raise errors.HeaderParseError(
                 "expected name-addr but found '{}'".format(token))
         if leader is not None:
-            token[0][:0] = [leader]
+            if isinstance(token[0], TokenList):
```
Comment (maxking) [inline Lib/email/_header_value_parser.py:1781]: This part of the fix doesn't seem to be related to the linked bpo, can you please explain the need for this?

I see that both the test cases are passing without this change.

File: Lib/email/_header_value_parser.py [hunk 2]
```diff
@@ -566,12 +566,12 @@ def display_name(self):
         if res[0].token_type == 'cfws':
             res.pop(0)
         else:
-            if res[0][0].token_type == 'cfws':
+            if isinstance(res[0][0], TokenList) and res[0][0].token_type == 'cfws':
```
Comment (maxking) [inline Lib/email/_header_value_parser.py]: You only need to check `res[0]` is a `TokenList` and not `rest[0][0]`. Same is true for the fix on line 574.

**Review by maxking — CHANGES_REQUESTED** [review]

File: Lib/email/_header_value_parser.py (hunk 1 above)
Comment (tsufeki) [reply Lib/email/_header_value_parser.py:1781]: The second case fails without this?
```
Run tests in parallel using 3 child processes
0:00:11 load avg: 2.76 [1/1/1] test_email failed
test test_email failed -- Traceback (most recent call last):
  File "/home/bb/tmp/cpython/Lib/email/_header_value_parser.py", line 1951, in get_address
    token, value = get_group(value)
  File "/home/bb/tmp/cpython/Lib/email/_header_value_parser.py", line 1909, in get_group
    raise errors.HeaderParseError("expected ':' at end of group "
email.errors.HeaderParseError: expected ':' at end of group display name but found '<jxd@example.com>'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/bb/tmp/cpython/Lib/test/test_email/test_headerregistry.py", line 1286, in test_complex_address_list
    h = self.make_header('to', source)
  File "/home/bb/tmp/cpython/Lib/test/test_email/test_headerregistry.py", line 79, in make_header
    return self.factory(name, value)
  File "/home/bb/tmp/cpython/Lib/email/headerregistry.py", line 602, in __call__
    return self[name](name, value)
  File "/home/bb/tmp/cpython/Lib/email/headerregistry.py", line 197, in __new__
    cls.parse(value, kwds)
  File "/home/bb/tmp/cpython/Lib/email/headerregistry.py", line 340, in parse
    kwds['parse_tree'] = address_list = cls.value_parser(value)
  File "/home/bb/tmp/cpython/Lib/email/headerregistry.py", line 331, in value_parser
    address_list, value = parser.get_address_list(value)
  File "/home/bb/tmp/cpython/Lib/email/_header_value_parser.py", line 1974, in get_address_list
    token, value = get_address(value)
  File "/home/bb/tmp/cpython/Lib/email/_header_value_parser.py", line 1954, in get_address
    token, value = get_mailbox(value)
  File "/home/bb/tmp/cpython/Lib/email/_header_value_parser.py", line 1777, in get_mailbox
    token, value = get_name_addr(value)
  File "/home/bb/tmp/cpython/Lib/email/_header_value_parser.py", line 1760, in get_name_addr
    token[0][:0] = [leader]
TypeError: 'ValueTerminal' object does not support item assignment

== Tests result: FAILURE ==

1 test failed:
    test_email

Total duration: 11 sec 830 ms
Tests result: FAILURE
```

**Comment by tsufeki** [comment]
I have made the requested changes; please review again

File: Lib/email/_header_value_parser.py (hunk 1 above)
Comment (maxking) [reply Lib/email/_header_value_parser.py:1781]: Ah, okay! Thanks!

File: Misc/NEWS.d/next/Library/2019-08-29-20-26-08.bpo-30988.b-_h5O.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix parsing invalid email address headers starting or ending with a dot.
```
Comment (maxking) [inline Misc/NEWS.d/next/Library/2019-08-29-20-26-08.bpo-30988.b-_h5O.rst]:
```suggestion
Fix parsing of emails with invalid address headers having a leading or trailing dot. Patch by tsufeki.
```

**Comment by malvidin** [comment]
@maxking and @csabella, is there any additional review that is needed for this pull request from @tsufeki? 
We would appreciate it if we could remove the workaround in the eml_parser for bpo-30988.
https://github.com/GOVCERT-LU/eml_parser/blob/f98980a77d9c7d914d97525a62294075c0ce42d9/tests/test_emlparser.py#L131

**Comment by spacefreak86** [comment]
Hi @maxking 

Could you please review this PL and approve it? This issue currently affects my productive systems and it would be great if it could be fixed upstream.

FYI: python/email-team was requested to review this PL, but it does not seem to exist anymore. Maybe that's the reason, why this PL was not approved yet?

File: Lib/email/_header_value_parser.py [hunk 4]
```diff
@@ -566,12 +566,12 @@ def display_name(self):
         if res[0].token_type == 'cfws':
             res.pop(0)
         else:
-            if res[0][0].token_type == 'cfws':
+            if isinstance(res[0], TokenList) and res[0][0].token_type == 'cfws':
```
Comment (ZackerySpytz) [inline Lib/email/_header_value_parser.py]: The PEP 8 style guide states that lines should be limited to 79 characters.

File: Lib/email/_header_value_parser.py [hunk 5]
```diff
@@ -566,12 +566,12 @@ def display_name(self):
...
                 res[0] = TokenList(res[0][1:])
         if res[-1].token_type == 'cfws':
             res.pop()
         else:
-            if res[-1][-1].token_type == 'cfws':
+            if isinstance(res[-1], TokenList) and res[-1][-1].token_type == 'cfws':
```
Comment (ZackerySpytz) [inline Lib/email/_header_value_parser.py]: Ditto.

File: Misc/NEWS.d/next/Library/2019-08-29-20-26-08.bpo-30988.b-_h5O.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Fix parsing of emails with invalid address headers having a leading or trailing dot. Patch by tsufeki.
```
Comment (ZackerySpytz) [inline Misc/NEWS.d/next/Library/2019-08-29-20-26-08.bpo-30988.b-_h5O.rst:1]: For what it's worth, the real name of the patch author is usually used in the news entry (though it's not mandatory).

**Comment by tsufeki** [comment]
@ZackerySpytz @spacefreak86 I fixed those long lines.

**Comment by malvidin** [comment]
Is there anyone from the python/email-team to review this? 
@ZackerySpytz, @maxking, and @csabella, do you know who can review this?

**Comment by spacefreak86** [comment]
@maxking could please review the changed PR again? It would be really great to have this fixed upstream.

**Comment by BoboTiG** [comment]
Gentle ping for this one since #2811 was closed :)

**Review by BoboTiG — APPROVED** [review]

**Comment by arhadthedev** [comment]
@warsaw (as a email module [expert](https://devguide.python.org/core-developers/experts/#stdlib))

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

There was also the same bug in `DisplayName.value`, which I have fixed.
